### PR TITLE
K2 FSA Reverse Implemenation

### PR DESCRIPTION
**Pull request by LvHang**
The Reverse() function is implemented for K2 FSAs.
I'm trying to test it.

### CODE REVIEW

File: k2/csrc/reverse.cu [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+    Array1<int32_t> reversed_row_split1 =
+      Plus(acc_non_empty_fsas, fsas_.RowSplits(1));
+    return RaggedShape2(&reversed_row_split1, nullptr, -1);
+  }
+
+
+  /*
+   Retrun the RaggedShape of the arc-level reversed FsaVec.
+   We generate the RaggedShape by figuring out the number of out-going arcs for
+   each reversed state. As the original source state will become to destination
```
Comment (csukuangfj) [inline k2/csrc/reverse.cu:141]:
```suggestion
   each reversed state. As the original source state will become destination
```

File: k2/csrc/reverse.cu [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+   state. So we collect the information by 'incoming_arcs'.
```
Comment (csukuangfj) [inline k2/csrc/reverse.cu:142]:
```suggestion
   state, we collect the information by 'incoming_arcs'.
```

File: k2/csrc/reverse.cu [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+    // doesn't have the out-going arcs. So we deal with the each original state.
+    K2_EVAL(
+        c_, num_ori_states, lambda_set_num_arcs_per_reversed_state, 
+        (int32_t idx01)->void {
+          int32_t idx0 = fsas_row_ids1_data[idx01],
+                  idx0x = fsas_row_splits1_data[idx0],
+                  num_arcs = (idx01 == idx0x ? 1 : 0)
+          num_arcs += incoming_arcs_row_splits2_data[idx01 + 1] -
+                      incoming_arcs_row_splits2_data[idx01];
+          num_arcs_per_reversed_state[old2new_map_[idx01]] = num_arcs;
```
Comment (csukuangfj) [inline k2/csrc/reverse.cu:171]:
```suggestion
          num_arcs_per_reversed_state_data[old2new_map_[idx01]] = num_arcs;
```

Also, please use `old2new_map_data[idx01]` in the kernel.
(You have to create `old2new_map_data` by yourself.

File: k2/csrc/reverse.cu [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+  RaggedShape CreateReversedArcSahpe(Ragged<int32_t> &incoming_arcs,
+                                     RaggedShape &reversed_states_shape) {
+    int32_t num_ori_states = fsas_.TotSize(1);
+    const int32_t *fsas_row_splits1_data = fsas_.RowSplits(1).Data(),
+                  *fsas_row_ids1_data = fsas_.RowIds(1).Data(), 
+                  *incoming_arcs_row_splits2_data =
+                    incoming_arcs.RowSplits(2).Data();
+
+    Array1<int32_t> num_arcs_per_reversed_state(
+        c_, reversed_states_shape.NumElements() + 1, 0);
```
Comment (csukuangfj) [inline k2/csrc/reverse.cu:157]: Do we have to initialize it to 0?

File: k2/csrc/reverse.cu [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+   2) For each non-empty original FSA, we add an extra "final_arc" which from
+      the 'penultimate' reversed state to the 'final' reversed state. It always
+      be the last arc for corresponding FSA.
+   */
+  Array1<Arc> ReverseArcsAndCreateArcMap(Ragged<int32_t> &incoming_arcs,
+                                         RaggedShape &reversed_shape
+                                         Array1<int32_t> *arc_map) {
+    int32_t num_reversed_arcs = reversed_shape.TotSize(2);
+    Array1<Arc> reversed_arcs(c_, num_reversed_arcs);
+    Arc *reversed_arcs_data =reversed_arcs.Data();
```
Comment (csukuangfj) [inline k2/csrc/reverse.cu:193]:
```suggestion
    Arc *reversed_arcs_data = reversed_arcs.Data();    
```
Please consider using `clang-format` to format your file.

File: k2/csrc/reverse_test.cu [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    arc_map = arc_map.To(cpu);
+    int32_t num_arcs = arcs.Dim();
+    for (int32_t i = 0; i != num_arcs; ++i) {
+      if (arc_map[i] != -1)
+        EXPECT_EQ(arcs[i].score, fsa_arcs[arc_map[i]].score);
+    }
+  }
+}
+
+TEST(Reverse, FsaVec) {
```
Comment (csukuangfj) [inline k2/csrc/reverse_test.cu:84]: Could you also add tests for empty FSAs?

File: k2/csrc/reverse.cu (hunk 4 above)
Comment (LvHang) [reply k2/csrc/reverse.cu:157]: After reversing, there is an extra new final state which has no outgoing arcs. So I initialize it to 0. Otherwise, I need another "K2_EVAL" to set it. So I choose a convenient way.

**Comment by LvHang** [comment]
The “EmptyFsaVec ” test has been added.
I fixed the bugs and solve your comments. Now, all the tests have passed.

**Comment by csukuangfj** [comment]
@LvHang Do you plan to also add the Python interface for `Reverse`?

**Comment by LvHang** [comment]
What should I do? I can try, it will help me to familiar with k2, I think. Is there anything I can refer to? Thanks

**Comment by danpovey** [comment]
@LvHang pick another function that has a unique name and that is wrapped to python, and grep for its name in the codebase.  You'll see what to do.

**Comment by danpovey** [comment]
> @LvHang pick another function that has a unique name and that is wrapped to python, and grep for its name in the codebase. You'll see what to do.  E.g. remove_epsilon_self_loops/RemoveEpsilonSelfLoops.

**Comment by LvHang** [comment]
OK, got it. Let me see, thanks.

**Comment by danpovey** [comment]
You can use `git grep <search_pattern>`

**Comment by LvHang** [comment]
The Python interface of reverse() has been added, and the "ctest --verbose -R reverse_test_py" has passed.
[I need to learn more things, like fsa_from_unary_function_tensor(), etc.]

File: k2/python/tests/reverse_test.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+#
+#  ctest --verbose -R reverse_test_py
+
+import unittest
+
+import k2
+import torch
+
+
+class TestConnect(unittest.TestCase):
```
Comment (csukuangfj) [inline k2/python/tests/reverse_test.py:30]:
```suggestion
class TestReverse(unittest.TestCase):
```

**Comment by csukuangfj** [comment]
> Perhaps this is ready to merge, @csukuangfj ?

Sorry. I missed this PR.

Looks good to me. Thanks @LvHang
